# ComplexomeProfiler

An application to discover insights from proteomics data by integrating with the complexome (available for multiple species). The core principle behind this software tool is that a more comprehensive understanding of functional perturbations can be achieved by contextualizing affected (i.e., differentially expressed) proteins within their complexes and functional assemblies in the cellular environment. Proteomics log fold change data (and associated statistical values) are used to analyze complexome remodeling between biological conditions (example, disease vs healthy control). Useful to understand disease phenotypes and identify (clinical) biomarkers.

Best used in <a href="https://colab.research.google.com/github/mguharoy/Complexome/blob/main/Complexome.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Google Colab"/></a>. See [tutorial](tutorial.md) for more information.

## Use cases

Typical use cases include:
- Sample vs control (e.g., disease condition, patient mutations)
- Perturbation (e.g., gene knockdown)-based phenotypic effects vs wild-type
- Estimate effects on pathway(s) and cross-talk by knockout/knockdown or chemically blocking pathways
- Trajectory (time t vs initial t0) studies
- etcâ€¦.

## Contributing

Pull requests are welcome. For major changes, please open an issue first
to discuss what you would like to change.

## License

[GPLv3](https://spdx.org/licenses/GPL-3.0-or-later.html)


## Citing

If you find this software useful, please cite:
[manuscript submitted] Add [CITATION.cff](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-citation-files)

Preprint describing the application of this software to metabolic disease is available at <https://www.biorxiv.org/content/10.1101/2023.12.30.573613v1>
